### PR TITLE
tutorial to use lightly to find new examples

### PR DESCRIPTION
**Pull request by MalteEbner**
closes https://github.com/lightly-ai/lightly-core/issues/939

## Description

- Creates a tutorial for predicting bounding boxes, cropping, saving objectness scores as metadata, uploading to webapp and using there. For more use case details see issue.
- [Tutorial 6: Find false negatives of object detection — lightly 1.0.0 documentation.pdf](https://github.com/lightly-ai/lightly/files/7014060/Tutorial.6.Find.false.negatives.of.object.detection.lightly.1.0.0.documentation.pdf)

- Refactors the cropping code to be usable not only for cropping from yolo labels, but doing cropping in general from any bounding boxes.
- Added unittests to new code parts.
- Added unittest to saving metadata

### CODE REVIEW

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    8. In the Lightly Webapp: Sort the images in the explore view by increasing objectness score. This allows to easily find missing examples / false positives and similar images to them.
+
+
+
+.. code-block:: python
+
+    import torch
+    import torchvision
+    from torch.utils.data import DataLoader
+    from torch.utils.hipify.hipify_python import bcolors
```
Comment (IgorSusmelj) [inline docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:66]: Let's keep this simple and remove any code not relevant for the tutorial. I would skip the `bcolors` and just do a normal print.

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        for cropped_images_filename, objectness_score in zip(cropped_images_list, objectness_scores_list):
+            metadata = {"objectness_score": float(objectness_score)}
+            metadata_list.append((cropped_images_filename, metadata))
+    save_custom_metadata(METADATA_OUTPUT_FILE, metadata_list)
+
+    ''' 6. Tell the lightly CLI command '''
+    cli_command = f"lightly-magic input_dir={OUTPUT_DIR} new_dataset_name=SKU_110k_val_cropped trainer.max_epochs=0 "
+    cli_command += f"custom_metadata={METADATA_OUTPUT_FILE} token=MY_TOKEN"
+    print(f"Upload the images and custom metadata with the following CLI command:")
+    print(f"{bcolors.OKBLUE}{cli_command}{bcolors.ENDC}")
```
Comment (IgorSusmelj) [inline docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:135]: drop `bcolors` here

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
```
Comment (IgorSusmelj) [inline docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:136]: Can we add a screenshot here on how the output of the terminal could look like? You can black out the token and dataset id using gimp

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    1. Define the dataset and torch dataloader for your dataset.
+    2. Define a pretrained object detection model. We use the retina net trained on COCO 2017. As it was not pretrained on a retail dataset, its performance is not state-of-the art. Nonetheless, it is sufficient for this tutorial and very easy to use.
+    3. Predict with the model on the dataset.
+    4. Use the bounding boxes of the object predictions to crop the objects out of the full images and save them in the output directory.
+    5. Extract the objectness scores and save them as custom metadata in a .json file.
+    6. Use the lightly-magic command to upload the cropped images, their embeddings and the objectness scores.
+    7. In the Lightly Webapp: Configure the objectness score as custom metadata.
+    8. In the Lightly Webapp: Sort the images in the explore view by increasing objectness score. This allows to easily find missing examples / false positives and similar images to them.
+
+
```
Comment (IgorSusmelj) [inline docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:59]: Can we add somewhere the expected dataset format? Should there be subfolders? Are these just images (e.g. png, jpg?)

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+* How to upload the cropped images with their corresponding metadata to the webapp.
+* How to use the webapp to find false negatives and similar examples easily.
+
+Requirements
+-------------
+
+You can use your own dataset or the one we provide for this tutorial. The dataset
+we will use is the `SKU110k dataset <https://github.com/eg4000/SKU110K_CVPR19>` showing store shelves.
+The tutorial is computationally expensive if you run it on the full 110k images of products,
+thus we recommend running it on a subset of the dataset.
```
Comment (IgorSusmelj) [inline docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:37]: Can you elaborate more on how you got the subset and how many images you ended up using? How long does it take to go through the tutorial? How did you download the dataset?

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+We directly see that many images show images despite having a low objectness score,
+thus they are false negatives / missing examples.
+
+.. figure:: ../../tutorials_source/platform/images/tutorial_cropped_objects_metadata/sku110k_explore_sort_objectness.png
+    :align: center
+    :alt: Explore view of images sorted by ascending objectness score
+
+When clicking on one them, we wee that they it has a low objectness score of only 0.05, despite showing an object,
+thus it is false negative. Similar images which are also false negatives are shown as well.
+Thus all of them can be added directly to the list of missing examples,
```
Comment (IgorSusmelj) [inline docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:158]: How can we add them to the list of missing examples? I think here we could also show a gif on how we add similar images

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py (hunk 4 above)
Comment (MalteEbner) [reply docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:59]: The dataset is just images. They can be in subfolders, but don't need to.

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py (hunk 6 above)
Comment (MalteEbner) [reply docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:158]: This feature is missing in the webapp.

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py (hunk 3 above)
Comment (MalteEbner) [reply docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:140]: I added one, but it does not look to good in my eyes.
If you don't disagree I would remove the screenshot again. It is only the usual output of a `lightly-magic` command, it would fit better to a general `lightly-magic` tutorial or documentation in my eyes.

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py (hunk 1 above)
Comment (MalteEbner) [reply docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py]: Done

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py (hunk 2 above)
Comment (MalteEbner) [reply docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py]: Done

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py (hunk 3 above)
Comment (MalteEbner) [reply docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:140]: I added a screenshot, but it does not look too good. I would remove it, I don't think it provides a lot of value, as it is just the usual lightly-magic output.

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+The tutorial is computationally expensive if you run it on the full 110k images of products,
+thus we recommend running it on a subset of the dataset. E.g. copy 100 images from one folder to
+a new folder and use the path to the latter folder as input. Alternatively, run it on your own dataset.
+
+For this tutorial the `lightly` pip package needs to be installed:
+
+.. code-block:: bash
+
+    # Install lightly as a pip package
+    pip install lightly.
```
Comment (IgorSusmelj) [inline docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:45]: Remove the `.` after lightly

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    3. Predict with the model on the dataset.
+    4. Use the bounding boxes of the object predictions to crop the objects out of the full images and save them in the output directory.
+    5. Extract the objectness scores and save them as custom metadata in a .json file.
+    6. Use the lightly-magic command to upload the cropped images, their embeddings and the objectness scores.
+    7. In the Lightly Webapp: Configure the objectness score as custom metadata.
+    8. In the Lightly Webapp: Sort the images in the explore view by increasing objectness score. This allows to easily find missing examples / false positives and similar images to them.
+
+Computational Expense
+-------------
+* Step 3 needs to run a model on every image in your dataset. In step 6, when embedding all cropped images, an embedding model needs to be run on every of them. These two are computationally expensive. They run much faster if you have CUDA support.
```
Comment (IgorSusmelj) [inline docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:62]: We don't have to go into details here about which steps require how much time. We also don't need a separate section for that.
I think we can keep it simple and just add a line somewhere: `This tutorials runs in about X min on a machine using a XYZ GPU with N GB of video memory`.

Comment (IgorSusmelj) [reply docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:62]: For me as a reader the important points are:
- Do I need a GPU or can I run it within 1-2 hours on a machine with no GPU at all?
- Do I need lots of memory? Is there a parameter I could tweak (e.g. batch size) if I run into issues?

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py (hunk 3 above)
Comment (IgorSusmelj) [reply docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:140]: I think the screenshot adds value but the text is too small. Any chance in doing the same with a smaller terminal window or using a zoomed terminal window?
Don't forget that this also acts as a check whether all the steps before were done correctly. If I run the tutorial and get a similar output I know that I'm on track.

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py (hunk 8 above)
Comment (MalteEbner) [reply docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:62]: In runs in <30mins on the CPU, so no need for that.

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py (hunk 3 above)
Comment (MalteEbner) [reply docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:140]: I added a screenshot with a bigger text.

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py (hunk 5 above)
Comment (MalteEbner) [reply docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py]: Also see https://github.com/lightly-ai/lightly/pull/462#discussion_r686684825

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py (hunk 6 above)
Comment (MalteEbner) [reply docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:164]: I added the screenshots + description.

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    # Install lightly as a pip package
+    pip install lightly
+
+Steps
+-------------
+
+The steps of this tutorial are quite straightforward:
+    1. Define the dataset in form of a LightlyDataset of torch tensors. You need to provide the path to the dataset images.
+    2. Define a pretrained object detection model. We use the retina net trained on COCO 2017. As it was not pretrained on a retail dataset, its performance is not state-of-the art. Nonetheless, it is sufficient for this tutorial and very easy to use.
+    3. Predict with the model on the dataset.
```
Comment (philippmwirth) [inline docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:53]: Is this a correct sentence? Why not "Generate predictions for each image in the dataset"

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    from torch.utils.data import DataLoader
+    from tqdm import tqdm
+
+    from lightly.active_learning.utils import BoundingBox
+    from lightly.data import LightlyDataset
+    from lightly.utils import save_custom_metadata
+    from lightly.utils.cropping.crop_image_by_bounding_boxes import crop_dataset_by_bounding_boxes_and_save
+
+    BASE_PATH = "path/to/dataset/"
+    DATASET_PATH = BASE_PATH+"images"  # the path were the full images are found
```
Comment (philippmwirth) [inline docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:75]: I would use `os.path.join` for all of these

**Review by philippmwirth — COMMENTED** [review]
I like the content of the tutorial, just added some feedback regarding wording.

File: docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py [hunk 11]
```diff
@@ -72,10 +72,10 @@
     from lightly.utils.cropping.crop_image_by_bounding_boxes import crop_dataset_by_bounding_boxes_and_save
 
     BASE_PATH = "path/to/dataset/"
-    DATASET_PATH = BASE_PATH+"images"  # the path were the full images are found
-    OUTPUT_DIR = BASE_PATH+"cropped_images" # the path where the cropped images will be saved
+    DATASET_PATH = os.path.join(BASE_PATH,"images")  # the path were the full images are found
```
Comment (philippmwirth) [inline docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:75]: Sorry for being so picky but you should always have a space after a comma.

Comment (MalteEbner) [reply docs/source/tutorials_source/platform/tutorial_cropped_objects_metadata.py:75]: That's write.
Usually my IDE tells me this, but it does not know that this is python code.

**Review by philippmwirth — APPROVED** [review]
